weblab/entities/views.py

```diff
@@ -721,3 +744,96 @@ def get_context_data(self, **kwargs):
+
+
+class EntityDiffView(View):
+    def _get_unix_diff(self, file1, file2):
+        with NamedTemporaryFile() as tmp1, NamedTemporaryFile() as tmp2:
+            tmp1.write(file1.data_stream.read())
+            tmp2.write(file2.data_stream.read())
+            tmp1.flush()
+            tmp2.flush()
+
+            result = {}
+            try:
+                output = subprocess.run(
+                    ['diff', '-a', tmp1.name, tmp2.name],
+                    stdout=subprocess.PIPE)
+                result['unixDiff'] = output.stdout.decode()
+                result['response'] = True
+            except subprocess.SubprocessError as e:
+                result['responseText'] = "Couldn't compute unix diff (%s)" % e
+
+            return result
+
+    def _get_bives_diff(self, file1, file2):
+        file1_url = 'file1_url'
+        file2_url = 'file2_url'
+        bives_url = settings.BIVES_URL 
+        post_data = {
+            'files': [
+                file1.data_stream.read().decode(),
+                file2.data_stream.read().decode(),
+            ],
+            'commands': [
+                'compHierarchyJson',
+                'reactionsJson',
+                'reportHtml',
+                'xmlDiff',
+            ]
+        }
+
+        result = {}
+        bives_response = requests.post(bives_url, json=post_data)
+
+        if bives_response.ok:
+            bives_json = bives_response.json()
+
+            if 'error' in bives_json:
+                result['responseText'] = '\n'.join(bives_json['error'])
+            else:
+                result['bivesDiff'] = bives_json
+                result['response'] = True
+        else:
+            result['responseText'] = (
+                'bives request failed: %d (%s)' %
+                (bives_response.status_code, bives_response.content.decode())
+            )
+
+        return result
+
+    def get(self, request, *args, **kwargs):
+        filename = self.kwargs['filename']
+        json_entities = []
+
+        versions = self.kwargs['versions'].strip('/').split('/')
+
+        diff_type = self.request.GET.get('type', 'unix')
+        if diff_type == 'unix':
+            task = 'getUnixDiff'
+        elif diff_type == 'bives':
+            task = 'getBivesDiff'
+
+        files = []
+        for version in versions:
+            id, sha = version.split(':')
+            try:
+                entity = Entity.objects.get(pk=id)
+                if entity.is_version_visible_to_user(sha, request.user):
+                    files.append(entity.repo.get_commit(sha).get_blob(filename))
+            except (RepoCacheMiss, Entity.DoesNotExist):
+                pass
+
+
+        if len(files) != 2:
+            result = {
+                'responseText': 'invalid request'
+            }
+        elif diff_type == 'unix':
+            result = self._get_unix_diff(*files)
+        elif diff_type == 'bives':
+            result = self._get_bives_diff(*files)
+
```

Comment:
We should add an `else` here providing an error message if `diff_type` has not been set correctly.